File: sdk/typescript/tests/openrouter_media_routing.test.ts
/**
 * Coverage for new OpenRouter routing / param-translation logic in
 * OpenRouterMediaProvider:
 *   - fetchModelMeta (HTTP success/cache/error/exception paths)
 *   - generateImage: imageUrls multi-part content, imageConfig snake_case mapping
 *   - generateVideo: imageUrl, frame_images snake_case, input_references, extra
 *   - generateAudio: /audio/speech path with speed + extra
 */
import {
  describe,
  it,
  expect,
  beforeEach,
  afterEach,
  vi,
} from 'vitest';
import { OpenRouterMediaProvider } from '../src/ai/OpenRouterMediaProvider.js';

const originalFetch = globalThis.fetch;
let mockFetch: ReturnType<typeof vi.fn>;

beforeEach(() => {
  mockFetch = vi.fn();
  globalThis.fetch = mockFetch;
});

afterEach(() => {
  globalThis.fetch = originalFetch;
});

describe('OpenRouterMediaProvider: model metadata fetch', () => {
  it('caches metadata after first GET', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({
        data: {
          id: 'openai/gpt-audio-mini',
          architecture: {
            output_modalities: ['text', 'audio'],
            input_modalities: ['text'],
          },
        },
      }),
    });
    // For both audio calls below — they hit chat-completions SSE.
    const sseBody = {
      getReader: () => {
        let done = false;
        return {
          read: async () => {
            if (done) return { done: true, value: undefined };
            done = true;
            const enc = new TextEncoder();
            return { done: false, value: enc.encode('data: [DONE]\n\n') };
          },
        };
      },
    };
    mockFetch.mockResolvedValueOnce({ ok: true, body: sseBody });
    mockFetch.mockResolvedValueOnce({ ok: true, body: sseBody });

    const provider = new OpenRouterMediaProvider({ apiKey: 'k' });
    await provider.generateAudio({
      text: 'a', model: 'openrouter/openai/gpt-audio-mini', format: 'mp3',
    });
    await provider.generateAudio({
      text: 'b', model: 'openrouter/openai/gpt-audio-mini', format: 'mp3',
    });
    // 1 metadata GET + 2 audio POSTs = 3 calls
    expect(mockFetch).toHaveBeenCalledTimes(3);
    const firstUrl = String(mockFetch.mock.calls[0][0]);
    expect(firstUrl).toContain('/models/openai/gpt-audio-mini/endpoints');
  });

  it('falls back to /audio/speech when metadata GET returns 500', async () => {
    mockFetch.mockResolvedValueOnce({ ok: false, status: 500 });
    mockFetch.mockResolvedValueOnce({
      ok: true,
      headers: { get: () => 'audio/mpeg' },
      arrayBuffer: async () => new Uint8Array([1, 2, 3]).buffer,
      text: async () => '',
    });

    const provider = new OpenRouterMediaProvider({ apiKey: 'k' });
    const resp = await provider.generateAudio({
      text: 'x', model: 'unknown/tts', format: 'mp3',
    });
    expect(resp.audio).not.toBeNull();
    expect(String(mockFetch.mock.calls[1][0])).toContain('/audio/speech');
  });

  it('falls back to /audio/speech when metadata GET throws', async () => {
    mockFetch.mockRejectedValueOnce(new Error('network'));
    mockFetch.mockResolvedValueOnce({
      ok: true,
      headers: { get: () => 'audio/mpeg' },
      arrayBuffer: async () => new Uint8Array([1]).buffer,
      text: async () => '',
    });
    const provider = new OpenRouterMediaProvider({ apiKey: 'k' });
    await provider.generateAudio({ text: 'x', model: 'unknown/y', format: 'mp3' });
    expect(String(mockFetch.mock.calls[1][0])).toContain('/audio/speech');
  });
});

describe('OpenRouterMediaProvider.generateVideo: param translation', () => {
  it('maps imageUrl, frameImages, inputReferences, extra to OpenRouter schema', async () => {
    // 1) submit, 2) poll → completed, 3) download
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({ id: 'jobZ' }),
    });
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({
        id: 'jobZ',
        status: 'completed',
        unsigned_urls: ['https://cdn.example.com/v.mp4'],
      }),
    });
    mockFetch.mockResolvedValueOnce({
      ok: true,
      arrayBuffer: async () => new Uint8Array([0x66, 0x66]).buffer,
    });

    const provider = new OpenRouterMediaProvider({ apiKey: 'k' });
    await provider.generateVideo({
      prompt: 'x',
      model: 'openrouter/google/veo-3.1-lite',
      duration: 4,
      imageUrl: 'https://example.com/seed.jpg',
      frameImages: [
        { imageUrl: { url: 'https://x/first.jpg' }, frameType: 'first_frame' },
        { imageUrl: { url: 'https://x/last.jpg' }, frameType: 'last_frame' },
      ],
      inputReferences: [{ imageUrl: { url: 'https://x/ref.jpg' } }],
      extra: { personGeneration: 'allow_all' },
      pollInterval: 1,
      timeout: 10_000,
    });

    const submitBody = JSON.parse(mockFetch.mock.calls[0][1].body as string);
    expect(submitBody.image_url).toBe('https://example.com/seed.jpg');
    expect(submitBody.frame_images).toEqual([
      {
        type: 'image_url',
        image_url: { url: 'https://x/first.jpg' },
        frame_type: 'first_frame',
      },
      {
        type: 'image_url',
        image_url: { url: 'https://x/last.jpg' },
        frame_type: 'last_frame',
      },
    ]);
    expect(submitBody.input_references).toEqual([
      { type: 'image_url', image_url: { url: 'https://x/ref.jpg' } },
    ]);
    expect(submitBody.personGeneration).toBe('allow_all');
  });
});

describe('OpenRouterMediaProvider.generateImage: imageUrls + imageConfig', () => {
  it('uses the current live Gemini image model by default', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({ choices: [] }),
    });

    const provider = new OpenRouterMediaProvider({ apiKey: 'k' });
    await provider.generateImage({ prompt: 'fox in watercolor' });

    const body = JSON.parse(mockFetch.mock.calls[0][1].body as string);
    expect(body.model).toBe('google/gemini-3.1-flash-image-preview');
  });

  it('builds multi-part user content and snake_cases imageConfig', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({
        choices: [
          {
            message: {
              content: null,
              images: [
                {
                  type: 'image_url',
                  image_url: { url: 'data:image/png;base64,QUJD' },
                },
              ],
            },
          },
        ],
      }),
    });

    const provider = new OpenRouterMediaProvider({ apiKey: 'k' });
    const resp = await provider.generateImage({
      prompt: 'fox in watercolor',
      model: 'openrouter/x-ai/grok-imagine-image-quality',
      imageUrls: ['https://r/1.png', 'https://r/2.png'],
      imageConfig: {
        aspectRatio: '16:9',
        imageSize: '1024x576',
        strength: 0.6,
        style: 'painterly',
        rgbColors: [[255, 0, 0]],
        backgroundRgbColor: [0, 0, 0],
        superResolutionReferences: ['https://r/sr.png'],
        fontInputs: [{ fontUrl: 'https://f.com/x.ttf', text: 'hi' }],
      },
      extra: { quality_boost: true },
    });

    const body = JSON.parse(mockFetch.mock.calls[0][1].body as string);
    // Multi-part content
    expect(body.messages[0].content).toEqual([
      { type: 'text', text: 'fox in watercolor' },
      { type: 'image_url', image_url: { url: 'https://r/1.png' } },
      { type: 'image_url', image_url: { url: 'https://r/2.png' } },
    ]);
    // imageConfig snake_case translation
    expect(body.image_config).toEqual({
      aspect_ratio: '16:9',
      image_size: '1024x576',
      strength: 0.6,
      style: 'painterly',
      rgb_colors: [[255, 0, 0]],
      background_rgb_color: [0, 0, 0],
      super_resolution_references: ['https://r/sr.png'],
      font_inputs: [{ font_url: 'https://f.com/x.ttf', text: 'hi' }],
    });
    expect(body.quality_boost).toBe(true);
    expect(body.modalities).toEqual(['image']);
    // Returned image base64 captured from data: URL
    expect(resp.images[0].b64Json).toBe('QUJD');
  });
});

describe('OpenRouterMediaProvider.generateAudio: /audio/speech extras', () => {
  it('uses the current live Kokoro model and voice by default', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      headers: { get: () => 'audio/mpeg' },
      arrayBuffer: async () => new Uint8Array([1, 2]).buffer,
      text: async () => '',
    });

    const provider = new OpenRouterMediaProvider({ apiKey: 'k' });
    provider.seedModelMeta('hexgrad/kokoro-82m', ['speech'], ['text']);
    await provider.generateAudio({ text: 'hello', format: 'mp3' });

    const body = JSON.parse(mockFetch.mock.calls[0][1].body as string);
    expect(body.model).toBe('hexgrad/kokoro-82m');
    expect(body.voice).toBe('af_alloy');
  });

  it('passes speed and extra through to the speech body', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      headers: { get: () => 'audio/mpeg' },
      arrayBuffer: async () => new Uint8Array([1, 2]).buffer,
      text: async () => '',
    });

    const provider = new OpenRouterMediaProvider({ apiKey: 'k' });
    provider.seedModelMeta('openai/gpt-4o-mini-tts', ['speech'], ['text']);
    await provider.generateAudio({
      text: 'bonjour',
      model: 'openai/gpt-4o-mini-tts',
      voice: 'alloy',
      format: 'mp3',
      speed: 1.5,
      extra: { language: 'fr-FR' },
    });
    const body = JSON.parse(mockFetch.mock.calls[0][1].body as string);
    expect(body.speed).toBe(1.5);
    expect(body.language).toBe('fr-FR');
    expect(body.response_format).toBe('mp3');
    expect(body.voice).toBe('alloy');
  });

  it('throws MediaProviderError when /audio/speech returns 401', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: false,
      status: 401,
      text: async () => 'unauthorized',
    });
    const provider = new OpenRouterMediaProvider({ apiKey: 'k' });
    provider.seedModelMeta('hexgrad/kokoro-82m', ['speech'], ['text']);
    await expect(
      provider.generateAudio({
        text: 'x', model: 'hexgrad/kokoro-82m', voice: 'af_bella', format: 'wav',
      })
    ).rejects.toThrow('Audio generation failed');
  });
});

File: sdk/go/ai/openrouter_media_routing_test.go
package ai

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// =============================================================================
// fetchModelMeta — happy path + cache reuse + network failures
// =============================================================================

func TestFetchModelMetaCachesResult(t *testing.T) {
	calls := 0
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		calls++
		assert.True(t, strings.HasSuffix(r.URL.Path, "/models/openai/gpt-audio-mini/endpoints"))
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"data":{"id":"openai/gpt-audio-mini","architecture":{"output_modalities":["text","audio"],"input_modalities":["text"]}}}`))
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: srv.URL, Client: srv.Client()}

	meta1 := p.fetchModelMeta(context.Background(), "openrouter/openai/gpt-audio-mini")
	meta2 := p.fetchModelMeta(context.Background(), "openai/gpt-audio-mini") // same model, no prefix

	assert.Equal(t, 1, calls, "cache should prevent second HTTP call")
	assert.Equal(t, []string{"text", "audio"}, meta1.OutputModalities)
	assert.Equal(t, meta1, meta2)
}

func TestFetchModelMetaReturnsEmptyOnHTTPError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusInternalServerError)
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: srv.URL, Client: srv.Client()}
	meta := p.fetchModelMeta(context.Background(), "unknown/model")
	assert.Empty(t, meta.OutputModalities)
	assert.Empty(t, meta.InputModalities)
}

func TestFetchModelMetaReturnsEmptyOnMalformedJSON(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`not-json`))
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: srv.URL, Client: srv.Client()}
	meta := p.fetchModelMeta(context.Background(), "x/y")
	assert.Empty(t, meta.OutputModalities)
}

func TestFetchModelMetaTriggersAutoRoutingViaSpeechEndpoint(t *testing.T) {
	// Server handles BOTH the metadata GET and the /audio/speech POST so the
	// provider can self-discover routing end-to-end.
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch {
		case strings.HasSuffix(r.URL.Path, "/models/hexgrad/kokoro-82m/endpoints"):
			w.Header().Set("Content-Type", "application/json")
			_, _ = w.Write([]byte(`{"data":{"id":"hexgrad/kokoro-82m","architecture":{"output_modalities":["speech"],"input_modalities":["text"]}}}`))
		case r.URL.Path == "/audio/speech":
			w.Header().Set("Content-Type", "audio/pcm")
			_, _ = w.Write(make([]byte, 240)) // raw PCM
		default:
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: srv.URL, Client: srv.Client()}
	resp, err := p.GenerateAudio(context.Background(), AudioRequest{
		Text:   "hi",
		Model:  "hexgrad/kokoro-82m",
		Voice:  "af_bella",
		Format: "wav",
	})
	require.NoError(t, err)
	require.NotNil(t, resp.Audio)
	decoded, err := base64.StdEncoding.DecodeString(resp.Audio.Data)
	require.NoError(t, err)
	assert.Equal(t, []byte("RIFF"), decoded[:4])
}

// =============================================================================
// generateAudioViaSpeechEndpoint — non-WAV format, error paths
// =============================================================================

func TestGenerateAudioSpeechMP3PassthroughNoWavWrap(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, "/audio/speech", r.URL.Path)
		var body map[string]any
		require.NoError(t, json.NewDecoder(r.Body).Decode(&body))
		assert.Equal(t, "mp3", body["response_format"])
		assert.Equal(t, 1.5, body["speed"])
		assert.Equal(t, "fr-FR", body["language"]) // from Extra
		w.Header().Set("Content-Type", "audio/mpeg")
		_, _ = w.Write([]byte("FAKE_MP3_BYTES_ABCDEF"))
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: srv.URL, Client: srv.Client()}
	p.SeedModelMeta("hexgrad/kokoro-82m", []string{"speech"}, []string{"text"})

	speed := 1.5
	resp, err := p.GenerateAudio(context.Background(), AudioRequest{
		Text:   "bonjour",
		Model:  "hexgrad/kokoro-82m",
		Voice:  "af_bella",
		Format: "mp3",
		Speed:  &speed,
		Extra:  map[string]any{"language": "fr-FR"},
	})
	require.NoError(t, err)
	require.NotNil(t, resp.Audio)
	assert.Equal(t, "mp3", resp.Audio.Format)
	// MP3 is returned as-is (no WAV wrap)
	decoded, err := base64.StdEncoding.DecodeString(resp.Audio.Data)
	require.NoError(t, err)
	assert.Equal(t, "FAKE_MP3_BYTES_ABCDEF", string(decoded))
}

func TestGenerateAudioSpeechErrorBubblesUp(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnauthorized)
		_, _ = w.Write([]byte(`{"error":"bad key"}`))
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: srv.URL, Client: srv.Client()}
	p.SeedModelMeta("hexgrad/kokoro-82m", []string{"speech"}, []string{"text"})

	_, err := p.GenerateAudio(context.Background(), AudioRequest{
		Text: "x", Model: "hexgrad/kokoro-82m", Voice: "af_bella", Format: "wav",
	})
	require.Error(t, err)
	assert.Contains(t, err.Error(), "audio/speech error")
	assert.Contains(t, err.Error(), "401")
}

func TestGenerateAudioSpeechDefaultsVoiceToAlloy(t *testing.T) {
	var seen map[string]any
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_ = json.NewDecoder(r.Body).Decode(&seen)
		w.Header().Set("Content-Type", "audio/mpeg")
		_, _ = w.Write([]byte("x"))
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: srv.URL, Client: srv.Client()}
	p.SeedModelMeta("openai/gpt-4o-mini-tts", []string{"speech"}, []string{"text"})
	_, err := p.GenerateAudio(context.Background(), AudioRequest{
		Text: "hi", Model: "openai/gpt-4o-mini-tts", Format: "mp3",
	})
	require.NoError(t, err)
	assert.Equal(t, "alloy", seen["voice"])
}

func TestGenerateAudioDefaultsToLiveKokoroModelAndVoice(t *testing.T) {
	var seen map[string]any
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		require.NoError(t, json.NewDecoder(r.Body).Decode(&seen))
		w.Header().Set("Content-Type", "audio/mpeg")
		_, _ = w.Write([]byte("x"))
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: srv.URL, Client: srv.Client()}
	p.SeedModelMeta("hexgrad/kokoro-82m", []string{"speech"}, []string{"text"})
	_, err := p.GenerateAudio(context.Background(), AudioRequest{
		Text: "hi", Format: "mp3",
	})
	require.NoError(t, err)
	assert.Equal(t, "hexgrad/kokoro-82m", seen["model"])
	assert.Equal(t, "af_alloy", seen["voice"])
}

// =============================================================================
// Video — first/last frame + auth-aware download
// =============================================================================

func TestGenerateVideoSendsFrameImagesAndImageURL(t *testing.T) {
	var submit map[string]any
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/videos":
			require.NoError(t, json.NewDecoder(r.Body).Decode(&submit))
			w.WriteHeader(http.StatusAccepted)
			_, _ = w.Write([]byte(`{"id":"jobABC"}`))
		case "/videos/jobABC":
			w.Header().Set("Content-Type", "application/json")
			_, _ = w.Write([]byte(`{
				"id":"jobABC","status":"completed",
				"unsigned_urls":["` + r.Host + `/blob/jobABC"],
				"usage":{"cost":0.42}
			}`))
		}
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: srv.URL, Client: srv.Client()}

	resp, err := p.GenerateVideo(context.Background(), VideoRequest{
		Prompt:   "test",
		Model:    "openrouter/google/veo-3.1-lite",
		Duration: 4,
		ImageURL: "https://example.com/seed.jpg",
		FrameImages: []map[string]any{
			{"type": "image_url", "image_url": map[string]string{"url": "https://x/first.jpg"}, "frame_type": "first_frame"},
			{"type": "image_url", "image_url": map[string]string{"url": "https://x/last.jpg"}, "frame_type": "last_frame"},
		},
		InputReferences: []map[string]any{
			{"type": "image_url", "image_url": map[string]string{"url": "https://x/ref.jpg"}},
		},
		Extra:        map[string]any{"personGeneration": "allow_all"},
		PollInterval: 10 * time.Millisecond,
		Timeout:      30 * time.Second,
	})
	require.NoError(t, err)
	require.NotNil(t, resp)
	require.Len(t, resp.Videos, 1)
	assert.Equal(t, 0.42, resp.Videos[0].CostUSD)

	// Verify submit payload included our params.
	assert.Equal(t, "google/veo-3.1-lite", submit["model"])
	assert.Equal(t, "https://example.com/seed.jpg", submit["image_url"])
	frames := submit["frame_images"].([]any)
	require.Len(t, frames, 2)
	assert.Equal(t, "first_frame", frames[0].(map[string]any)["frame_type"])
	assert.Equal(t, "last_frame", frames[1].(map[string]any)["frame_type"])
	assert.Equal(t, "allow_all", submit["personGeneration"])
}

func TestGenerateVideoDownloadsWithoutAuthFromNonOpenRouterHost(t *testing.T) {
	// Use a fixed URL string captured after server start so the poll response
	// can embed an absolute http:// URL.
	var apiSrv *httptest.Server
	apiSrv = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/videos":
			w.WriteHeader(http.StatusAccepted)
			_, _ = w.Write([]byte(`{"id":"job1"}`))
		case "/videos/job1":
			w.Header().Set("Content-Type", "application/json")
			_, _ = w.Write([]byte(fmt.Sprintf(
				`{"id":"job1","status":"completed","unsigned_urls":["%s/blob/job1"],"usage":{"cost":0.1}}`,
				apiSrv.URL)))
		case "/blob/job1":
			// Non-openrouter host → no Authorization header.
			assert.Empty(t, r.Header.Get("Authorization"))
			w.Header().Set("Content-Type", "video/mp4")
			_, _ = w.Write([]byte("FAKE_MP4"))
		}
	}))
	defer apiSrv.Close()

	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: apiSrv.URL, Client: apiSrv.Client()}
	resp, err := p.GenerateVideo(context.Background(), VideoRequest{
		Prompt: "t", Model: "x/y",
		PollInterval: 10 * time.Millisecond, Timeout: 30 * time.Second,
	})
	require.NoError(t, err)
	require.Len(t, resp.Videos, 1)
	decoded, err := base64.StdEncoding.DecodeString(resp.Videos[0].Data)
	require.NoError(t, err)
	assert.Equal(t, "FAKE_MP4", string(decoded))
}

// =============================================================================
// Image — ImageURLs reference images + image_config conversion
// =============================================================================

func TestGenerateImageWithReferenceImagesAndConfig(t *testing.T) {
	var captured map[string]any
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		require.NoError(t, json.NewDecoder(r.Body).Decode(&captured))
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"choices":[{"message":{"content":null,"images":[{"type":"image_url","image_url":{"url":"data:image/png;base64,QUJD"}}]}}]}`))
	}))
	defer srv.Close()

	strength := 0.7
	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: srv.URL, Client: srv.Client()}
	resp, err := p.GenerateImage(context.Background(), ImageRequest{
		Prompt:    "a fox in watercolor",
		Model:     "openrouter/x-ai/grok-imagine-image-quality",
		ImageURLs: []string{"https://x/ref1.png", "https://x/ref2.png"},
		ImageConfig: &ImageConfig{
			AspectRatio: "16:9",
			Strength:    &strength,
			RgbColors:   [][3]int{{255, 100, 50}},
		},
		Extra: map[string]any{"high_quality": true},
	})
	require.NoError(t, err)
	require.Len(t, resp.Images, 1)
	assert.Equal(t, "QUJD", resp.Images[0].B64JSON)

	// Modalities sent as ["image"] only.
	mods := captured["modalities"].([]any)
	require.Len(t, mods, 1)
	assert.Equal(t, "image", mods[0])

	// User content is a multi-part array (text + 2 image_url entries).
	messages := captured["messages"].([]any)
	userMsg := messages[0].(map[string]any)
	content := userMsg["content"].([]any)
	require.Len(t, content, 3)
	assert.Equal(t, "text", content[0].(map[string]any)["type"])
	assert.Equal(t, "image_url", content[1].(map[string]any)["type"])
	assert.Equal(t, "image_url", content[2].(map[string]any)["type"])

	// Extra passthrough.
	assert.Equal(t, true, captured["high_quality"])
}

// =============================================================================
// wrapPCM16AsWAV — header correctness
// =============================================================================

func TestWrapPCM16AsWAVHeader(t *testing.T) {
	pcm := make([]byte, 240) // 240 bytes = 120 16-bit samples = 5ms @ 24kHz
	wav := wrapPCM16AsWAV(pcm, 24000)

	assert.Equal(t, []byte("RIFF"), wav[:4])
	assert.Equal(t, []byte("WAVE"), wav[8:12])
	assert.Equal(t, []byte("fmt "), wav[12:16])
	assert.Equal(t, []byte("data"), wav[36:40])
	// total file size = 44 (header) + 240 (data)
	assert.Len(t, wav, 44+240)
}

// =============================================================================
// SeedModelMeta — chat-audio path verification on a separate model
// =============================================================================

func TestSeedModelMetaRoutesToChatCompletions(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, "/chat/completions", r.URL.Path)
		w.Header().Set("Content-Type", "text/event-stream")
		flusher, _ := w.(http.Flusher)
		fmt.Fprint(w, "data: [DONE]\n\n")
		flusher.Flush()
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{APIKey: "k", BaseURL: srv.URL, Client: srv.Client()}
	p.SeedModelMeta("openai/gpt-audio-mini", []string{"text", "audio"}, []string{"text"})

	_, err := p.GenerateAudio(context.Background(), AudioRequest{
		Text: "hi", Model: "openai/gpt-audio-mini", Format: "mp3",
	})
	require.NoError(t, err)
}
